python/datasources/age_adjust.py

```diff
@@ -0,0 +1,204 @@
+        int_cols = [std_col.COVID_CASES, std_col.COVID_HOSP_Y,
+                    std_col.COVID_HOSP_N, std_col.COVID_HOSP_UNKNOWN,
+                    std_col.COVID_DEATH_Y, std_col.COVID_DEATH_N,
+                    std_col.COVID_DEATH_UNKNOWN]
+
+        for table_name, df in table_names_to_dfs.items():
+            # All columns are str, except outcome columns.
+            column_types = {c: 'STRING' for c in df.columns}
+            for col in int_cols:
+                if col in column_types:
+                    column_types[col] = 'FLOAT'
+            if std_col.RACE_INCLUDES_HISPANIC_COL in df.columns:
+                column_types[std_col.RACE_INCLUDES_HISPANIC_COL] = 'BOOL'
+
+            # Clean up column names.
+            self.clean_frame_column_names(df)
+
+            gcs_to_bq_util.add_dataframe_to_bq(
+                df, dataset, table_name, column_types=column_types)
+
+
+def merge_age_adjusted(df, age_adjusted_df):
+    """Merges the age adjusted death rate into the standard COVID dataset.
+       Returns a dataframe with all needed COVID info for the frontend.
+
+       df: a dataframe with covid date without age adjusted numbers
+       age_adjusted_df: a dataframe with age adjusted covid numbers"""
+
+    merge_cols = [std_col.STATE_FIPS_COL, std_col.STATE_NAME_COL]
+    merge_cols.extend(std_col.RACE_COLUMNS)
+
+    df = df.reset_index(drop=True)
+    age_adjusted_df = age_adjusted_df.reset_index(drop=True)
+
+    return pd.merge(df, age_adjusted_df, how='left', on=merge_cols)
+
+
+def get_expected_deaths_and_hospitalizations(race_and_age_df, population_df):
+    """Calculates the age adjusted expected deaths of each racial group.
+       I made this function to break up the age adjustment into smaller, more easily testable pieces.
+       Returns a dataframe meant to be used in memory.
+
+       race_and_age_df: a dataframe with covid deaths broken down by race and age
+       population_df: a dataframe with population broken down by race and age"""
+
+    def get_expected_rate(row):
+        this_pop_size = float(population_df.loc[
+                (population_df[std_col.RACE_CATEGORY_ID_COL] == row[std_col.RACE_CATEGORY_ID_COL]) &
+                (population_df[std_col.AGE_COL] == row[std_col.AGE_COL]) &
+                (population_df[std_col.STATE_FIPS_COL] == row[std_col.STATE_FIPS_COL])
+            ][std_col.POPULATION_COL].values[0])
+
+        ref_pop_size = float(population_df.loc[
+                (population_df[std_col.RACE_CATEGORY_ID_COL] == REFERENCE_POPULATION) &
+                (population_df[std_col.AGE_COL] == row[std_col.AGE_COL]) &
+                (population_df[std_col.STATE_FIPS_COL] == row[std_col.STATE_FIPS_COL])
+            ][std_col.POPULATION_COL].values[0])
+
+        if not row[std_col.COVID_DEATH_Y]:
+            row['expected_deaths'] = None
+        else:
+            true_death_rate = float(row[std_col.COVID_DEATH_Y]) / this_pop_size
```

Comment:
what happens if `this_pop_size` is zero / undefined; will division by zero choke this? Probably it's fine but wasn't sure if we should add some guards against that

Reply:
the line `if not row[std_col.COVID_DEATH_Y]:` will guard against that. We know we have a population number for every demographic that has a covid count. 

It can't hurt to throw an error though if you think that would be helpful

Reply:
I'm adding the error check